### PR TITLE
Added SharedCompilationId property to ManagedCompiler and inherited tasks

### PR DESCRIPTION
**Pull request by zahess**
**Customer scenario**

Allows specification of a user-defined shared compilation ID to _replace_ standard pipe name construction for VBCSCompiler.exe. This feature exists when calling csc.exe or vbc.exe directly, but it cannot be specified via MSBuild task. This new task property mirrors the behavior of csc/vbc.

**Bugs this fixes:**

https://mseng.visualstudio.com/VSOnline/_workitems?id=1011824

**Workarounds, if any**

It could be possible to modify build tasks to execute csc/vbc directly with a **/shared** switch, and specify that switch value with a project property, but this is unnecessarily complicated, may impact performance, and may run into other blocking issues depending on build requirements.

**Risk**

As behavior mirrors existing csc/vbc behavior there are some risks which are also shared, specifically:
- csc/vbc accept a value to their **/shared** switch and do not perform any error checking on it
- ManagedCompiler likewise does not perform input validation on SharedCompilationId
- In both above cases, the provided value entirely _replaces_ construction of the pipe name for VBCSCompiler.
- This value is also used as a Mutex name when an instance of VBCSCompiler is registered; illegal characters in the pipe name will lead to failures.

Additionally, specification of a **SharedCompilationId** property value for a compile task while **UseSharedCompilation** value of false should be considered invalid input. The current behavior is to silently ignore the **SharedCompilationId** value entirely.

**Performance impact**

Nominal impact; ManagedCompiler tasks simply pass the SharedCompilationId value through to BuildServerConnection.RunServerCompilation (with a single string.IsNullOrEmpty check). Build server connection checks the value for null before following existing behavior. There actually may be a small performance improvement when SharedCompilationId is specified as standard pipe name construction is skipped if it is specified.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
Test windows_debug_unit64_prtest please

**Comment by AlekseyTs** [comment]
Test windows_release_unit64_prtest please

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review.

File: src/Compilers/Core/MSBuildTask/ManagedCompiler.cs [hunk 1]
```diff
@@ -288,6 +288,12 @@ public ITaskItem[] ResponseFiles
             get { return (ITaskItem[])_store[nameof(ResponseFiles)]; }
         }
 
+        public string SharedCompilationId
```
Comment (jcouv) [inline src/Compilers/Core/MSBuildTask/ManagedCompiler.cs]: The only thing that itches me with this PR is whether it needs some documentation. The Roslyn repo has a document for command-line arguments, but none for the Csc/Vbc tasks.

On the other hand, the docs repo has some. 
Here's the page: https://docs.microsoft.com/en-us/visualstudio/msbuild/csc-task
Here's the repo: https://github.com/dotnet/docs/blob/be7974018ce3195dc7344192d647fe64fb2ebcc4/xml/Microsoft.Build.Tasks/Csc.xml

Could you file an issue there to track updating that page once 15.5 is released? You can link it to this PR.

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for a second review. Thanks

File: src/Compilers/Core/MSBuildTask/ManagedCompiler.cs [hunk 2]
```diff
@@ -463,6 +469,7 @@ protected override int ExecuteTool(string pathToTool, string responseFileCommand
 
                     var responseTask = BuildServerConnection.RunServerCompilation(
                         Language,
+                        SharedCompilationId == "" ? null : SharedCompilationId,
```
Comment (jaredpar) [inline src/Compilers/Core/MSBuildTask/ManagedCompiler.cs]: Lets use `string.IsNullOrEmpty(SharedCompilationId)` to be safe here.

**Comment by jaredpar** [comment]
Tests?

**Comment by agocke** [comment]
Agree with @jaredpar, I would expect to see some tests added to the CscTests file and VbcTests file.

**Pull request merged**